### ISSUE TITLE
Preempted agents permacancel builds

### ISSUE THREAD

**Issue by benesch** [JetBrains/teamcity-google-agent#16]
When an agent is preempted, any build that it was running is permanently canceled:

![screenshot 2018-01-23 10 43 42](https://user-images.githubusercontent.com/882976/35285018-4baffc8a-002a-11e8-8c37-c4d6f16b3708.png)

I've noticed that TeamCity can automatically retry a build when someone runs `agent.sh stop force` on the agent, so I tried specifying the following shutdown script in the instance metadata:

```
{
  "shutdown-script": "sudo -u agent /home/agent/bin/agent.sh stop force"
}
```

Unfortunately, this doesn't actually work. I'm not sure whether it's GCE failing to run the startup script, or the cloud plugin noticing that the instance has been shutdown and somehow marking the build as permanently canceled. Either way, it would be really nice if TC could automatically restart a build that failed because the agent was preempted/removed.

Perhaps there's some way to do this already that I'm not aware of?

**Comment by dtretyakov** [comment]
@benesch, by default (without shoutdown script) TeamCity will restart this build on a new agent with a following reason:
![image](https://user-images.githubusercontent.com/1025927/35328710-88c701da-00fd-11e8-85a5-0c7e9bdb34c9.png)

So, actually you don't need to use some specific scripts for graceful TeamCity build agent shutdown. If the cancelled build is a part of build chain it also should be re-scheduled as a new build in chain. Do you see another behavior in case of cancellation with the reason `Canceled with comment: Agent removed`? If so could you please share the build dependency settings for dependent configuration.

**Comment by dtretyakov** [comment]
@benesch, did you managed to make it work as expected?

**Comment by supersteves** [comment]
For me, it works fine. Builds restart as described (unexpected finish).

**Comment by supersteves** [comment]
Nope, I spoke too soon. I'm seeing the same, occasionally. I've seen both cases. Sometimes builds stop as @dtretyakov says, and restart properly. Other times the build shows *Agent removed* as per @benesch. I've just seen it happen once, causing all dependant builds to fail also.

**Comment by supersteves** [comment]
For example: 
```
Result:	 Canceled (Tests passed: 806, ignored: 7; exit code 143 (new))
Canceled with comment: unexpectedFinish
Agent:	 visokio-teamcity-agent-linux-cloud-4-8-pd-std-15 (<Non existing pool> pool)
Time:	26 Mar 18 10:35 - 10:43 (7m:21s)	
Process exited with code 143 
```
Logs (teamcity-cloud):
```
[2018-03-26 08:54:20,619]   INFO [ing Executor 22] - ces.terminate.AgentRemoverImpl - Removing cloud agent "visokio-teamcity-agent-linux-cloud-4-8-pd-std-15" {id=591} because: Cloud instance (id=visokio-teamcity-agent-linux-cloud-4-8-pd-std-15, imageId=visokio-teamcity-agent-linux-cloud-4-8-pd-std-, profileId=google-7) has gone (is not reported by cloud profile) 
[2018-03-26 10:34:43,468]   INFO [3 Flush Queue 1] - uds.base.AbstractCloudInstance - Changing visokio-teamcity-agent-linux-cloud-4-8-pd-std-15(4f56a72c) status from Unknown to Scheduled to start  
[2018-03-26 10:34:43,468]   INFO [3 Flush Queue 1] - .server.impl.CloudEventsLogger - Cloud instance entered 'scheduled to start' state, profile 'Google Compute Engine cloud, linux'{id=google-7, projectId=project3}, GoogleCloudInstance{myName='visokio-teamcity-agent-linux-cloud-4-8-pd-std-15'} 
[2018-03-26 10:34:43,468]   INFO [3 Flush Queue 1] - .server.impl.CloudEventsLogger - Cloud instance start succeeded: profile 'Google Compute Engine cloud, linux'{id=google-7, projectId=project3}, GoogleCloudInstance{myName='visokio-teamcity-agent-linux-cloud-4-8-pd-std-15'} 
[2018-03-26 10:34:56,291]   INFO [onPool-worker-0] - clouds.google.GoogleCloudImage - Creating new virtual machine visokio-teamcity-agent-linux-cloud-4-8-pd-std-15 
[2018-03-26 10:35:09,686]   INFO [onPool-worker-3] - uds.base.AbstractCloudInstance - Changing visokio-teamcity-agent-linux-cloud-4-8-pd-std-15(4f56a72c) status from Scheduled to start to Running  
[2018-03-26 10:35:11,306]   INFO [uled executor 3] - .server.impl.CloudEventsLogger - Cloud instance entered 'running' state, profile 'Google Compute Engine cloud, linux'{id=google-7, projectId=project3}, GoogleCloudInstance{myName='visokio-teamcity-agent-linux-cloud-4-8-pd-std-15'} 
[2018-03-26 10:35:55,083]   INFO [io-8111-exec-38] - .server.impl.CloudEventsLogger - Detected cloud agent "visokio-teamcity-agent-linux-cloud-4-8-pd-std-15" {id=627}, profile 'Google Compute Engine cloud, linux'{id=google-7, projectId=project3}, GoogleCloudInstance{myName='visokio-teamcity-agent-linux-cloud-4-8-pd-std-15'} 
[2018-03-26 10:44:41,344]   INFO [{id=google-7} 2] - base.tasks.UpdateInstancesTask - Updated instance 'visokio-teamcity-agent-linux-cloud-4-8-pd-std-15' status to Stopped based on API information 
[2018-03-26 10:44:41,344]   INFO [{id=google-7} 2] - uds.base.AbstractCloudInstance - Changing visokio-teamcity-agent-linux-cloud-4-8-pd-std-15(4f56a72c) status from Running to Stopped  
[2018-03-26 10:44:41,344]   INFO [onPool-worker-2] - nnector.GoogleApiConnectorImpl - Removing terminated instance visokio-teamcity-agent-linux-cloud-4-8-pd-std-15 
[2018-03-26 10:44:46,333]   INFO [uled executor 5] - .server.impl.CloudEventsLogger - Cloud instance entered 'stopped' state, profile 'Google Compute Engine cloud, linux'{id=google-7, projectId=project3}, GoogleCloudInstance{myName='visokio-teamcity-agent-linux-cloud-4-8-pd-std-15'} 
[2018-03-26 10:44:46,334]   INFO [uled executor 5] - .server.impl.CloudEventsLogger - Cloud agent Unregistered agent "visokio-teamcity-agent-linux-cloud-4-8-pd-std-15" {id=627} is no longer recognized as cloud agent because instance has gone: profileId=google-7, imageId=visokio-teamcity-agent-linux-cloud-4-8-pd-std-, instanceId=visokio-teamcity-agent-linux-cloud-4-8-pd-std-15 
[2018-03-26 10:44:46,334]   INFO [uled executor 5] - ces.terminate.AgentRemoverImpl - Removing cloud agent Unregistered agent "visokio-teamcity-agent-linux-cloud-4-8-pd-std-15" {id=627} because: Cloud instance (id=visokio-teamcity-agent-linux-cloud-4-8-pd-std-15, imageId=visokio-teamcity-agent-linux-cloud-4-8-pd-std-, profileId=google-7) has gone (is not reported by cloud profile) 
```
Dependant build (composite):
```
Result:	 Build chain finished (success: 5, cancelled: 1); Tests passed: 20190, ignored: 71	
Time:	26 Mar 18 10:29 - 11:03 (34m:06s)	
Snapshot dependency failed  … Parallel :: Tests, unit [Java] Q3  Feature/TeamcityCloudCompatibility  #tc2902-f1e6448fd165 Canceled (Tests passed: 806, ignored: 7; exit code 143 (new)) 
 71 tests ignored
```
Seen in teamcity-server log:
```
[2018-03-26 10:29:44,930]   WARN - ldServer.AGENT.PollingProtocol - Error while executing 'runBuild' command on Agent '"visokio-teamcity-agent-linux-cloud-4-8-pd-std-5" {id=622}': jetbrains.buildServer.serverSide.impl.agent.AgentTi
meoutException: The agent didn't respond within the timeout period (60 seconds). (enable debug to see stacktrace) 
[2018-03-26 10:29:44,932]   WARN -   jetbrains.buildServer.SERVER - Error while starting build Omniscope 3.0 and up :: 1 Pass :: Compile Java {id=Omniscope3plus_CompilationJavaVcsRollUp, internal id=bt95} #tc31244-f1e6448fd165 {build id=2092000, promotion id=2092000, branch=Feature/TeamcityCloudCompatibility, history=false, agent=Unregistered agent "visokio-teamcity-agent-linux-cloud-4-8-pd-std-5" {id=622}, triggered by "'steve' (Steve Spencer) {id=8}" (##userId='8' type='snapshotDependency' buildTypeId='bt101' buildId='2092002'), canceled=[comment: Agent removed, userId:null]. Started 2018-03-26 10:28:44.897, finished <null>. Status "UNKNOWN 'Canceled'"}: jetbrains.buildServer.serverSide.impl.agent.AgentTimeoutException: The agent didn't respond within the timeout period (60 seconds). (enable debug to see stacktrace) 
```
Let me know if I should post this on the main Teamcity issue tracker.

**Comment by supersteves** [comment]
Cross-posted: https://youtrack.jetbrains.com/issue/TW-54336

**Comment by dtretyakov** [comment]
The [problem in TeamCity core](https://youtrack.jetbrains.com/issue/TW-54336) was resolved in TeamCity 2018.1.1. Please let us know if it still remains after upgrade.
